File: packages/typescript/examples/10_webhooks.ts
/**
 * Webhooks and push notifications.
 *
 * Run: npx ts-node examples/10_webhooks.ts
 */
import { Agent } from '../src';

const agent = new Agent({ name: 'WebhookDemo', description: 'Shows webhooks' });

agent.onComplete((skill, result) => {
  console.log('✅ Skill "' + skill + '" completed: ' + JSON.stringify(result));
});

agent.skill('long_task', { description: 'Task that takes time' }, async ({ duration = 2, callback_url }: { duration?: number; callback_url?: string }) => {
  console.log('Starting task (' + duration + 's)...');
  await new Promise(resolve => setTimeout(resolve, duration * 1000));
  const result = { status: 'completed', duration };
  if (callback_url) {
    try {
      await fetch(callback_url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(result),
      });
    } catch (e) { console.error('Webhook failed'); }
  }
  return result;
});

agent.skill('process_with_progress', { description: 'Task with progress' }, async ({ items = 10, callback_url }: { items?: number; callback_url?: string }) => {
  for (let i = 0; i < items; i++) {
    await new Promise(resolve => setTimeout(resolve, 500));
    if (callback_url) {
      try {
        await fetch(callback_url, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ progress: (i + 1) / items }),
        });
      } catch {}
    }
  }
  return { processed: items, status: 'done' };
});

agent.run({ port: 8787 });

File: packages/java/examples/10_Webhooks.java
import com.a2alite.Agent;
import com.a2alite.SkillConfig;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Map;

/**
 * Webhooks and push notifications.
 *
 * Run: ./gradlew run -PmainClass=Webhooks
 */
class Webhooks {
    private static final HttpClient httpClient = HttpClient.newHttpClient();
    private static final ObjectMapper mapper = new ObjectMapper();

    public static void main(String[] args) {
        var agent = Agent.builder().name("WebhookDemo").description("Shows webhooks").build();

        agent.onComplete((skill, result) -> System.out.printf("✅ Skill '%s' completed: %s%n", skill, result));

        agent.skill("long_task", SkillConfig.of("Task that takes time"), params -> {
            double duration = ((Number) params.getOrDefault("duration", 2.0)).doubleValue();
            String callbackUrl = (String) params.get("callback_url");
            System.out.printf("Starting task (%ss)...%n", duration);
            try { Thread.sleep((long) (duration * 1000)); }
            catch (InterruptedException e) { Thread.currentThread().interrupt(); }
            var result = Map.of("status", "completed", "duration", duration);
            if (callbackUrl != null) notifyWebhook(callbackUrl, result);
            return result;
        });

        agent.skill("process_with_progress", SkillConfig.of("Task with progress"), params -> {
            int items = ((Number) params.getOrDefault("items", 10)).intValue();
            String callbackUrl = (String) params.get("callback_url");
            for (int i = 0; i < items; i++) {
                try { Thread.sleep(500); } catch (InterruptedException e) { Thread.currentThread().interrupt(); }
                if (callbackUrl != null) notifyWebhook(callbackUrl, Map.of("progress", (i+1.0)/items));
            }
            return Map.of("processed", items, "status", "done");
        });

        agent.run(8787);
    }

    private static void notifyWebhook(String url, Object payload) {
        try {
            var request = HttpRequest.newBuilder().uri(URI.create(url))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(mapper.writeValueAsString(payload))).build();
            httpClient.sendAsync(request, HttpResponse.BodyHandlers.discarding());
        } catch (Exception e) { System.err.println("Webhook failed: " + e.getMessage()); }
    }
}
